expr/func_linearregression.go

```diff
@@ -0,0 +1,166 @@
+
+	return context
+}
+
+func (s *FuncLinearRegression) parseSourceAt() error {
+	loc, err := time.LoadLocation("")
+	if err != nil {
+		return fmt.Errorf("failed to load location: %w", err)
+	}
+	now := time.Now()
+
+	defaultStartSource := uint32(now.Add(-24 * time.Hour).Unix())
+	s.startSource, err = dur.ParseDateTime(s.startSourceAt, loc, now, defaultStartSource)
+	if err != nil {
+		return fmt.Errorf("failed to parse 'startSourceAt' argument %q: %w", s.startSourceAt, err)
+	}
+
+	defaultEndSource := uint32(now.Unix())
+	s.endSource, err = dur.ParseDateTime(s.endSourceAt, loc, now, defaultEndSource)
+	if err != nil {
+		return fmt.Errorf("failed to parse 'endSourceAt' argument %q: %w", s.endSourceAt, err)
+	}
+
+	return nil
+}
+
+func (s *FuncLinearRegression) Exec(dataMap DataMap) ([]models.Series, error) {
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	results := []models.Series{}
+	for _, serie := range series {
+		factor, offset, isValid := linearRegressionAnalysis(serie)
+		if !isValid {
+			continue
+		}
+
+		normalizedStartTarget := normalize(s.startTarget, serie.Interval)
+		size := int((s.endTarget-normalizedStartTarget)/serie.Interval + 1)
+		datapoints := pointSlicePool.GetMin(size)
+		for i := 0; i < size; i++ {
+			datapoint := schema.Point{
+				Val: offset + (float64(normalizedStartTarget)+float64(i)*float64(serie.Interval))*factor,
+				Ts:  normalizedStartTarget + uint32(i)*serie.Interval,
+			}
+			datapoints = append(datapoints, datapoint)
+		}
+
+		newSeries := serie.Copy([]schema.Point{})
+		newSeries.Target = fmt.Sprintf("linearRegression(%s, %d, %d)", serie.Target, s.startSource, s.endSource)
+		newSeries.Datapoints = datapoints
+		newSeries.Tags["linearRegressions"] = fmt.Sprintf("%d, %d", s.startSource, s.endSource)
+		newSeries.QueryPatt = newSeries.Target
+		newSeries.QueryFrom = s.startTarget
+		newSeries.QueryTo = s.endTarget
+
+		results = append(results, newSeries)
+	}
+
+	dataMap.Add(Req{}, results...)
+	return results, nil
+}
+
+func linearRegressionAnalysis(series models.Series) (float64, float64, bool) {
+	// Some functions normalize the series' datapoint timestamps,
+	// but not the series' QueryFrom/QueryTo fields.
+	startSource := series.QueryFrom
+	if len(series.Datapoints) > 0 {
+		startSource = series.Datapoints[0].Ts
+	}
+
+	var n float64
+	var sumI float64
+	var sumII float64
+	var sumV float64
+	var sumIV float64
+	for i, point := range series.Datapoints {
+		if math.IsNaN(point.Val) {
+			continue
+		}
+
+		n++
+		sumI += float64(i)
+		sumII += float64(i) * float64(i)
+		sumV += point.Val
+		sumIV += float64(i) * point.Val
+	}
+
+	denominator := n*sumII - sumI*sumI
+	if denominator == 0 {
+		return 0, 0, false
+	}
+	factor := (n*sumIV - sumI*sumV) / denominator / float64(series.Interval)
+	offset := (sumII*sumV-sumIV*sumI)/denominator - factor*float64(startSource)
+
+	return factor, offset, true
+}
+
+func normalize(timestamp, interval uint32) uint32 {
```

Comment:
Looks similar to [AggBoundary](https://github.com/grafana/metrictank/blob/cd09d04d4c784beecf3978016cda6695747c003c/mdata/aggregator.go#L11)